evap/results/tools.py

```diff
@@ -171,42 +170,102 @@ def _calculate_results_impl(course):
+        grade_questions_average = avg([result.average for result in grade_question_results])
+        grade_questions_distribution = avg_distribution([result.distribution for result in grade_question_results])
+        non_grade_question_results = [result for result in grouped_results[contributor] if result.question.is_non_grade_rating_question]
+        non_grade_questions_average = avg([result.average for result in non_grade_question_results])
+        non_grade_questions_distribution = avg_distribution([result.distribution for result in non_grade_question_results])
+        if contributor:
+            # collect the grade question average, non-grade question average and the distributions for each contributor
+            contributor_averages.append(mix(grade_questions_average, non_grade_questions_average, settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE))
+            contributor_distributions.append(weighted_avg_distribution(
+                [grade_questions_distribution, non_grade_questions_distribution],
+                [settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE, 1 - settings.CONTRIBUTOR_GRADE_QUESTIONS_PERCENTAGE]
+            ))
+        else:
+            # calculate the grade question average, non-grade question average and the distributions for the course
+            course_grade_questions_average = grade_questions_average
+            course_grade_questions_distribution = grade_questions_distribution
+            course_non_grade_questions_average = non_grade_questions_average
+            course_non_grade_questions_distribution = non_grade_questions_distribution
+
+    # calculate the average grade and distribution over all contributors
+    contributors_average = avg(contributor_averages)
+    contributors_distribution = avg_distribution(contributor_distributions)
+
+    # calculate to which percentage each group influences the total grade
+    contributions_percentage = settings.CONTRIBUTIONS_PERCENTAGE if contributors_average is not None else 0
+    course_grade_questions_percentage = settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_grade_questions_average is not None else 0
+    course_non_grade_questions_percentage = 1 - settings.CONTRIBUTIONS_PERCENTAGE - settings.COURSE_GRADE_QUESTIONS_PERCENTAGE if course_non_grade_questions_average is not None else 0
+
+    scale = 1 / sum([contributions_percentage, course_grade_questions_percentage, course_non_grade_questions_percentage])
+    contributions_percentage *= scale
+    course_grade_questions_percentage *= scale
+    course_non_grade_questions_percentage *= scale
+
+    # calculate average grade
+    average_grade = contributions_percentage * (contributors_average or 0) + \
+                    course_grade_questions_percentage * (course_grade_questions_average or 0) + \
+                    course_non_grade_questions_percentage * (course_non_grade_questions_average or 0)
```

Comment:
1.+5. avg function with optional weights for floats and distributions it is.

2. i would say using `None`s in distributions makes things more difficult because we would have to replace them with `0`s if other values are valid before returning them.

3. percentage calculation is now removed.

4. initialization changed again.

Reply:
2. well, the same goes for the averages doesn't it? so whatever we do for the averages we could do for the distributions. or maybe i'm missing something.